File: source-connector/src/main/java/sourceconnector/config/OffsetManagerConfiguration.java
package sourceconnector.config;

import lombok.extern.slf4j.Slf4j;
import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import sourceconnector.domain.connect.OffsetRecordService;
import sourceconnector.repository.offset.HttpOffsetRecordRepository;
import sourceconnector.repository.offset.InternalOffsetRecordRepository;
import sourceconnector.service.offset.OffsetRecordRepository;
import sourceconnector.service.offset.OffsetRecordServiceImpl;

import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;

@Slf4j
@Configuration
public class OffsetManagerConfiguration {

  @Bean
  public OffsetRecordRepository offsetRecordRepository(
    OffsetManagerConfig offsetManagerConfig,
    KafkaConsumer<String, Long> kafkaConsumer,
    AdminClient  adminClient,
    TopicConfig topicConfig
  ) {

    String baseUrl = offsetManagerConfig.baseUrl();
    if (baseUrl == null || baseUrl.isEmpty()) {
      log.info("baseUrl is omitted, {} is to be used.", InternalOffsetRecordRepository.class.getSimpleName());
      return new InternalOffsetRecordRepository(kafkaConsumer, adminClient, topicConfig.offsetTopic());
    }

    try {
      URL url = URI.create(offsetManagerConfig.baseUrl()).toURL();
      String protocol = url.getProtocol();

      switch (protocol) {
        case "http", "https" -> {
          return new HttpOffsetRecordRepository(url.toString());
        }
        default -> throw new IllegalArgumentException("Unsupported protocol: " + protocol);
      }
    } catch (MalformedURLException e) {
      throw new IllegalArgumentException("Invalid baseUrl: " + baseUrl);
    }
  }

  @Bean
  public OffsetRecordService offsetRecordService(OffsetRecordRepository offsetRecordRepository) {
    return new OffsetRecordServiceImpl(offsetRecordRepository);
  }
}

File: source-connector/src/main/kotlin/sourceconnector/config/OffsetManagerConfiguration.kt
package sourceconnector.config

import org.apache.kafka.clients.admin.AdminClient
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.slf4j.LoggerFactory
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import sourceconnector.config.OffsetManagerConfig.RepositoryType
import sourceconnector.config.OffsetManagerConfig.RepositoryType.HTTP
import sourceconnector.config.OffsetManagerConfig.RepositoryType.INTERNAL
import sourceconnector.domain.connect.OffsetRecordService
import sourceconnector.repository.offset.HttpOffsetRecordRepository
import sourceconnector.repository.offset.InternalOffsetRecordRepository
import sourceconnector.service.offset.OffsetRecordRepository
import sourceconnector.service.offset.OffsetRecordServiceImpl
import java.net.URI
import java.net.URL

@Configuration
class OffsetManagerConfiguration {
  private val log = LoggerFactory.getLogger(OffsetManagerConfiguration::class.java)
  @Bean
  fun offsetRecordRepository(
    offsetManagerConfig: OffsetManagerConfig,
    kafkaConsumer: KafkaConsumer<String, Long>,
    adminClient: AdminClient,
    topicConfig: TopicConfig
  ): OffsetRecordRepository {
    val type: RepositoryType = offsetManagerConfig.type
    when (type) {
      INTERNAL -> {
        log.info("Using InternalOffsetRecordRepository as offset record repository.")
        return InternalOffsetRecordRepository(kafkaConsumer, adminClient, topicConfig.offsetTopic)
      }

      HTTP -> {
        val baseUrl = offsetManagerConfig.baseUrl
        if (baseUrl.isNullOrEmpty()) {
          throw IllegalArgumentException("baseUrl must be provided for HTTP repository type.")
        }
        if (!baseUrl.startsWith("http://") && !baseUrl.startsWith("https://")) {
          throw IllegalArgumentException("Invalid baseUrl: $baseUrl")
        }

        return HttpOffsetRecordRepository(offsetManagerConfig.baseUrl)
      }
    }
  }


  @Bean
  fun offsetRecordService(offsetRecordRepository: OffsetRecordRepository): OffsetRecordService {
    return OffsetRecordServiceImpl(offsetRecordRepository)
  }
}
